flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/ProcessingTimeoutTrigger.java

```diff
@@ -0,0 +1,100 @@
+import org.apache.flink.api.common.typeutils.base.LongSerializer;
+import org.apache.flink.streaming.api.windowing.windows.Window;
+
+/**
+ * A {@link Trigger} that can turn any {@link Trigger} into a timeout {@code Trigger}.
+ *
+ * <p>Each record arriving will emit a ProcessingTimeTimer withing the interval,
+ * you can control if the timer will be registered periodic for each event arriving,
+ * by the continualInterval flag.
+ * you can control if the state will be cleared after reach the timeout, by the
+ * shouldClearAtTimeout flag.
+ *
+ * @param <T> The type of elements on which this trigger can operate.
+ * @param <W> The type of {@link Window Windows} on which this trigger can operate.
+ */
+
+@PublicEvolving
+public class ProcessingTimeoutTrigger<T, W extends Window> extends Trigger<T, W> {
+
+	private static final long serialVersionUID = 1L;
+
+	private Trigger<T, W> nestedTrigger;
+	private final long interval;
+	private final boolean continualInterval;
+	private final boolean shouldClearAtTimeout;
+
+	private final ValueStateDescriptor<Long> timeoutStateDesc;
+
+	private ProcessingTimeoutTrigger(Trigger<T, W> nestedTrigger, long interval, boolean continualInterval, boolean shouldClearAtTimeout) {
+		this.nestedTrigger = nestedTrigger;
+		this.interval = interval;
+		this.continualInterval = continualInterval;
+		this.shouldClearAtTimeout = shouldClearAtTimeout;
+		this.timeoutStateDesc = new ValueStateDescriptor<Long>("timeout", LongSerializer.INSTANCE);
+	}
+
+	@Override
+	public TriggerResult onElement(T element, long timestamp, W window, TriggerContext ctx) throws Exception {
+		TriggerResult triggerResult = this.nestedTrigger.onElement(element, timestamp, window, ctx);
+		if (triggerResult.isFire()) {
+			this.clear(window, ctx);
+			return triggerResult;
+		}
+		
+		ValueState<Long> timeoutState = ctx.getPartitionedState(this.timeoutStateDesc);
+		long nextFireTimestamp = ctx.getCurrentProcessingTime() + this.interval;
+
+		if (timeoutState.value() != null && continualInterval) {
+			ctx.deleteProcessingTimeTimer(timeoutState.value());
+			timeoutState.clear();
+		}
+
+		if (timeoutState.value() == null) {
+			timeoutState.update(nextFireTimestamp);
+			ctx.registerProcessingTimeTimer(nextFireTimestamp);
+		}
+
+		return triggerResult;
+	}
+
+	@Override
+	public TriggerResult onProcessingTime(long timestamp, W window, TriggerContext ctx) throws Exception {
+		this.nestedTrigger.onProcessingTime(timestamp, window, ctx);
+		if (shouldClearAtTimeout) {
+			this.clear(window, ctx);
+		}
+		return TriggerResult.FIRE;
+	}
+
+	@Override
+	public TriggerResult onEventTime(long timestamp, W window, TriggerContext ctx) throws Exception {
+		this.nestedTrigger.onEventTime(timestamp, window, ctx);
+		return TriggerResult.FIRE;
+	}
+
+	@Override
+	public void clear(W window, TriggerContext ctx) throws Exception {
+		ctx.getPartitionedState(this.timeoutStateDesc).clear();
+		this.nestedTrigger.clear(window, ctx);
```

Comment:
A trigger should also clear any potential timers on `clear()`. See for example the `EventTimeTrigger` or `ProcessingTimeTrigger`